aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/entities/EntityImpl.java

```diff
@@ -197,7 +198,7 @@ protected EntityImpl(AWSXRayRecorder creator, String name) {
 
         this.creator = creator;
         this.name = name;
-        this.subsegments = new ArrayList<>();
+        this.subsegments = new CopyOnWriteArrayList<>();
```

Comment:
This seems like it could get very expensive for people adding lots of subsegments... I know we don't want to spend the time for the most performant solution, but this seems like a pretty substantial degradation that would impact nearly all customers for more of an edge-case deadlock. Is there any other workaround for this?

Reply:
I think for something fairly standard like 5 subsegments, this is probably faster than something like a synchronized list.

For a hundred segments it does look slow - but so does the normal `ArrayList` since we call `remove` on it which results in quadratic behavior anyways.

I think to be optimal we would need to use a concurrent hash set, but unfortunately we need to support `getSubsegments().add()`. It is possible to wrap a set into a list (with slow performances of `get(I)`) I think, it could be an idea for if there are reports of regression in the wild.

Reply:
Ok gotcha, I guess we're also still calling `remove` so there's going to definitely be some degradation there. Can we add a quick additional benchmark test for creating say 100 subsegments just to confirm it's not breaking anything?